### DISCUSSION TITLE
Fall Guys crashing even after installing EOS and enabling Proton EAC runtime (Solution: LD_PRELOAD needs to be set)

### DISCUSSION THREAD

**Discussion by TheDcoder** [legendary-gl/legendary#456] [Q&A] [answered]
Hi,

I'm trying to run Fall Guys on Linux with my DIY setup, I don't want to use Lutris or the Hero Launcher (bloated frontend for Legendary? no thanks).

I've had success with my setup until I tried to join a game, it promptly informed that Easy Anti-Cheat wasn't enabled. So I looked into it all and got it to enable with the custom GE wine build and the EAC runtime which was downloaded by Lutris (bonus points if someone can figure out where this can be downloaded without Lutris).

Launched the game and still didn't work, but found out that it was looking for the EOS overlay, so I ran the installation command and that fixed that issue!

But now I can't get past another annoying error: (output from wine)
```
FWindowsPlatform::TryGetValueFromRegistry: Failed to get value from registry with SubKeyName 'SOFTWARE\Epic Games\EOS\UserHelper' and ValueName 'Path'
FWindowsBootstrapper::TryGetUserHelperPath: Path does not exist in registry
FGenericBootstrapper::Bootstrap: Failed to find path to User Helper
```

It seems EOS was not installed properly and the game is trying to read the location of the "User Helper". So what went wrong? Legendary still thinks EOS is installed:
```
$ legendary eos-overlay --prefix $WINEPREFIX info
[cli] INFO: Using prefix "/mnt/data_partition/Storage/Arch Linux/Wine/.wine_proton_ge"
[cli] INFO: Installed version: 1.2.5++EOSSDK+Release-Overlay-CL-20622577-Windows
[cli] INFO: Installed path: /mnt/data_partition/Storage/Arch Linux/Wine/Games/Epic Games/.overlay
[cli] INFO: Found available Overlay installations in:
[cli] INFO:  - /mnt/data_partition/Storage/Arch Linux/Wine/Games/Epic Games/.overlay
[cli] INFO: Overlay enabled: Yes
[cli] INFO: Enabled Overlay path: /mnt/data_partition/Storage/Arch Linux/Wine/Games/Epic Games/.overlay
[cli] INFO: No enabled Vulkan layers.
```

How do I fix this? Can anyone who is able to run Fall Guys check what's in this registry key?

My config:
```
[0a2d9f6403244d12969e11da6713137b]
wine_executable = /mnt/data/Storage/Arch Linux/Wine/Proton/GE/lutris-GE-Proton7.22/bin/wine64

[0a2d9f6403244d12969e11da6713137b.env]
PROTON_EAC_RUNTIME = /home/wineuser/.local/share/lutris/runtime/eac_runtime
WINEPREFIX = /mnt/data_partition/Storage/Arch Linux/Wine/.wine_proton_ge
```

Full wine output:
```
wineserver: using server-side synchronization.
wine: RLIMIT_NICE is <= 20, unable to use setpriority safely
002c:fixme:winediag:LdrInitializeThunk wine-staging 7.0 is a testing version containing experimental patches.
002c:fixme:winediag:LdrInitializeThunk Please mention your exact version when filing bug reports on winehq.org.
0078:err:ntoskrnl:ZwLoadDriver failed to create driver L"\\Registry\\Machine\\System\\CurrentControlSet\\Services\\wineusb": c0000142
003c:fixme:service:scmdatabase_autostart_services Auto-start service L"wineusb" failed to start: 1114
MESA-INTEL: warning: Performance support disabled, consider sysctl dev.i915.perf_stream_paranoid=0
MESA-INTEL: warning: Performance support disabled, consider sysctl dev.i915.perf_stream_paranoid=0
MESA-INTEL: warning: Performance support disabled, consider sysctl dev.i915.perf_stream_paranoid=0
0070:err:xrandr:xrandr14_get_adapters Failed to get adapters
FGenericBootstrapper::Bootstrap: Started Epic Online Services Bootstrapper 1.0.0++EOSSDK+Release-FGS6-CL-18019720
FGenericBootstrapper::Bootstrap: Getting config path
FGenericBootstrapper::Bootstrap: Reading config from path: 'D:\Games\Epic Games\FallGuys\FallGuys_client.ini'
FGenericBootstrapper::Bootstrap: Creating session
FGenericBootstrapper::TryCreateSession: Creating process group
FGenericBootstrapper::Bootstrap: Getting path to User Helper
FWindowsPlatform::TryGetValueFromRegistry: Failed to get value from registry with SubKeyName 'SOFTWARE\Epic Games\EOS\UserHelper' and ValueName 'Path'
FWindowsBootstrapper::TryGetUserHelperPath: Path does not exist in registry
FGenericBootstrapper::Bootstrap: Failed to find path to User Helper
FGenericBootstrapper::Bootstrap: Executing target application: 'D:\Games\Epic Games\FallGuys\FallGuys_client_game.exe'
FGenericBootstrapper::Bootstrap: Succeeded
0150:fixme:thread:NtSetInformationThread Can't set other thread's platform description
014c:fixme:thread:NtSetInformationThread Can't set other thread's platform description
0148:fixme:thread:NtSetInformationThread Can't set other thread's platform description
0154:fixme:thread:NtSetInformationThread Can't set other thread's platform description
0158:fixme:thread:NtSetInformationThread Can't set other thread's platform description
015c:fixme:thread:NtSetInformationThread Can't set other thread's platform description
0160:fixme:thread:NtSetInformationThread Can't set other thread's platform description
0164:fixme:thread:NtSetInformationThread Can't set other thread's platform description
0168:fixme:thread:NtSetInformationThread Can't set other thread's platform description
016c:fixme:thread:NtSetInformationThread Can't set other thread's platform description
0170:fixme:thread:NtSetInformationThread Can't set other thread's platform description
0174:fixme:thread:NtSetInformationThread Can't set other thread's platform description
0178:fixme:thread:NtSetInformationThread Can't set other thread's platform description
017c:fixme:thread:NtSetInformationThread Can't set other thread's platform description
0180:fixme:thread:NtSetInformationThread Can't set other thread's platform description
0184:fixme:thread:NtSetInformationThread Can't set other thread's platform description
0188:fixme:thread:NtSetInformationThread Can't set other thread's platform description
018c:fixme:thread:NtSetInformationThread Can't set other thread's platform description
0190:fixme:thread:NtSetInformationThread Can't set other thread's platform description
0194:fixme:thread:NtSetInformationThread Can't set other thread's platform description
0134:fixme:ntdll:NtQuerySystemInformation info_class SYSTEM_PERFORMANCE_INFORMATION
0198:fixme:thread:NtSetInformationThread Can't set other thread's platform description
019c:fixme:thread:NtSetInformationThread Can't set other thread's platform description
01a0:fixme:thread:NtSetInformationThread Can't set other thread's platform description
01a4:fixme:thread:NtSetInformationThread Can't set other thread's platform description
01a8:fixme:thread:NtSetInformationThread Can't set other thread's platform description
0134:err:wldap32:DllMain No libldap support, expect problems
0134:err:kerberos:kerberos_LsaApInitializePackage no Kerberos support, expect problems
0134:fixme:imm:ImmReleaseContext (000000000002005A, 000000006A15BD50): stub
MESA-INTEL: warning: Performance support disabled, consider sysctl dev.i915.perf_stream_paranoid=0
0134:fixme:wbemprox:client_security_SetBlanket 00000002419C3DA0, 0000000069FAFD50, 10, 0, (null), 3, 3, 0000000000000000, 0x00000000
0134:fixme:wbemprox:client_security_Release 00000002419C3DA0
01e4:fixme:thread:NtSetInformationThread Can't set other thread's platform description
0134:fixme:wbemprox:client_security_SetBlanket 00000002419C3DA0, 000000007EA214D0, 10, 0, (null), 3, 3, 0000000000000000, 0x00000000
0134:fixme:wbemprox:client_security_Release 00000002419C3DA0
0134:fixme:wbemprox:wbem_services_CreateInstanceEnum unsupported flags 0x00000030
0134:fixme:wbemprox:wbem_services_CreateInstanceEnum unsupported flags 0x00000030
020c:fixme:thread:NtSetInformationThread Can't set other thread's platform description
0220:fixme:thread:NtSetInformationThread Can't set other thread's platform description
0134:fixme:wbemprox:client_security_SetBlanket 00000002419C3DA0, 000000007EA28DD0, 10, 0, (null), 3, 3, 0000000000000000, 0x00000000
0134:fixme:wbemprox:client_security_Release 00000002419C3DA0
0134:fixme:wbemprox:wbem_services_CreateInstanceEnum unsupported flags 0x00000030
0134:fixme:wbemprox:wbem_services_CreateInstanceEnum unsupported flags 0x00000030
0134:fixme:wbemprox:wbem_services_CreateInstanceEnum unsupported flags 0x00000030
0224:fixme:thread:NtSetInformationThread Can't set other thread's platform description
0070:fixme:imm:ImeSetActiveContext (0x4803d0, 0): stub
0070:fixme:imm:ImmReleaseContext (0000000000010020, 00000000004803D0): stub
0228:fixme:thread:NtSetInformationThread Can't set other thread's platform description
terminate called after throwing an instance of 'std::logic_error'
  what():  basic_string::_S_construct null not valid
013c:fixme:file:CancelSynchronousIo (0000000000000080): stub
013c:fixme:kernelbase:AppPolicyGetProcessTerminationMethod FFFFFFFFFFFFFFFA, 000000000011FEB0
0230:fixme:kernelbase:AppPolicyGetProcessTerminationMethod FFFFFFFFFFFFFFFA, 000000000011FEB0
```

**Comment by CommandMC** [comment, marked as answer]
The overlay is not the issue, this is:
```
terminate called after throwing an instance of 'std::logic_error'
  what():  basic_string::_S_construct null not valid
```
Make sure to set `LD_PRELOAD` to *some value* before launching the game. The EAC runtime assumes it's set (since it is automatically set in Steam) and crashes if it isn't

Just one of those things found out by (& automatically fixed in) your friendly "bloated frontend for Legendary"

---

> bonus points if someone can figure out where this can be downloaded without Lutris

https://lutris.net/api/runtimes, "eac_runtime" there

**Reply by TheDcoder** [reply]
Thanks for the swift response, and also thanks for linking to the Lutris API end-point, useful stuff!

However I'm not so sure about your `LD_PRELOAD` trick, it needs to be a path to a valid `.so`, otherwise the runtime will exit even before the real application is started, here are some tests that I did:
```
$ export LD_PRELOAD=/dev/null
$ /usr/bin/false
ERROR: ld.so: object '/dev/null' from LD_PRELOAD cannot be preloaded (file too short): ignored.
$ export LD_PRELOAD=dummy
$ /usr/bin/false
ERROR: ld.so: object 'dummy' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
```
> Just one of those things found out by (& automatically fixed in) your friendly "bloated frontend for Legendary"

Can you provide the source for this bit of information? Just want to cross-check what exactly they are doing.

**Reply by CommandMC** [reply]
Trying to load an invalid .so file will fail, yes, but the runtime should still start

> Can you provide the source for this bit of information? Just want to cross-check what exactly they are doing.

We're setting LD_PRELOAD to an empty value if it isn't set yet. [Relevant code](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/blob/3608cc1c814f5ef9399e485df7f7b9ff74cfc1cc/electron/launcher.ts#L259-L264)

**Reply by TheDcoder** [reply]
You are right, had a dumb moment there and didn't even realize the output said 'ignored'. I also found out that setting it to an empty value did the trick and the game worked! No idea how the code to get `LD_PRELOAD` without safety is in such a critical section :dizzy_face:

Even played a full match to make sure that everything worked :)

> [Relevant code](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/blob/3608cc1c814f5ef9399e485df7f7b9ff74cfc1cc/electron/launcher.ts#L259-L264)

Thanks, that's very useful. Didn't realize that you were one of the contributors! No offense with the whole thing about the launcher being bloated, it's just that I don't like a whole browser to run my games. It's a fine tool for the not-so-technically inclined and for Windows users.

---

Aside from all that, my game was running like an ass with frequent frame drops compared to the experience I had in Windows on the same PC. Is it because I shifted it from the SSD to a regular spinny disk or is it due to the user-space EAC?

**Reply by CommandMC** [reply]
> Aside from all that, my game was running like an ass with frequent frame drops compared to the experience I had in Windows on the same PC. Is it because I shifted it from the SSD to a regular spinny disk or is it due to the user-space EAC?

Make sure DXVK is installed. Do not try to enable ESYNC/FSYNC, I've heard that the game crashes randomly with those enabled

**Reply by TheDcoder** [reply]
DXVK is OK, I've disable {E,F}SYNC. But the performance is still the same.

I'm highly suspecting the EAC because the game runs at ~60 FPS during the beginning of the match where the camera pans over the entire map, but it drops to below 1 FPS as soon as the match starts and players start jumping around... And perhaps this is just confirmation bias but I've feel like it lags a lot more when the players are jumping, this is very noticeable in levels where the task is to keep jumping.

Is this the anti-cheat doing rigorous calculations to prevent flying cheats?

**Reply by CommandMC** [reply]
Hm, I'm not quite sure, I didn't encounter any issues like this

Shot in the dark, but maybe Wine is logging a lot of stuff and your terminal is having problems keeping up? Could you try adding `WINEDEBUG = -all` to the environment variables?

Other than that, what system specifications are we looking at here? `inxi -SCGm` would be a good command to run to get all kinds of info

**Reply by TheDcoder** [reply]
> Shot in the dark, but maybe Wine is logging a lot of stuff and your terminal is having problems keeping up?

Nope, the output is actually very minimal compared to other games, it doesn't even output enough to make the terminal scroll more than once.

> Other than that, what system specifications are we looking at here?
```
System:
  Host: arch Kernel: 5.18.11-xanmod1-1 arch: x86_64 bits: 64 Desktop: Xfce
    v: 4.16.0 Distro: Arch Linux
Memory:
  RAM: total: 15.52 GiB used: 8.03 GiB (51.7%)
  RAM Report:
    permissions: Unable to run dmidecode. Root privileges required.
CPU:
  Info: quad core model: Intel Core i7-7700HQ bits: 64 type: MT MCP cache:
    L2: 1024 KiB
  Speed (MHz): avg: 3578 min/max: 800/3800 cores: 1: 3676 2: 3525 3: 3616
    4: 3584 5: 3593 6: 3562 7: 3679 8: 3393
Graphics:
  Device-1: Intel HD Graphics 630 driver: i915 v: kernel
  Device-2: NVIDIA GP107M [GeForce GTX 1050 Ti Mobile] driver: nvidia
    v: 515.57
  Device-3: IMC Networks USB2.0 UVC HD Webcam type: USB driver: uvcvideo
  Display: x11 server: X.Org v: 21.1.4 with: Xwayland v: 22.1.3 driver: X:
    loaded: modesetting,nvidia gpu: i915 resolution: 1: 1920x1080~60Hz 2: N/A
  OpenGL: renderer: Mesa Intel HD Graphics 630 (KBL GT2) v: 4.6 Mesa 22.1.3
```

**Reply by CommandMC** [reply]
Ah, a laptop, fun stuff
How are you making sure the game is using your Nvidia GPU?

**Reply by TheDcoder** [reply]
By using some special env variables which make the OpenGL special sauce choose Nvidia. I also verified that the Nvidia GPU is being used by looking at DXVK's HUD feature :)

**Reply by CommandMC** [reply]
Hm, welp, got nothing then. Everything's working fine here

**Reply by TheDcoder** [reply]
Do you mind sharing your specs? Perhaps the slowdown is also due to the recent software mitigations that had to be done to protect against CPU exploits. Most "gaming" kernels have these disabled so maybe that's the reason I'm getting FPS drops. I will test this theory tomorrow.

**Reply by TheDcoder** [reply]
Turning off all mitigations with the kernel parameter `mitigations=off` did result in a very significant improvement in performance! I can play the game now, there are places where it still lags but it does not sustain for more than a second, often less than that.

One interesting observation I made are the fat lag spikes on the home screen for Fall Guys, not quite sure what's up with that.
